analyzer/evmverifier/evmverifier.go

```diff
@@ -0,0 +1,183 @@
+package evmverifier
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ethCommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/evmverifier/sourcify"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/queries"
+	"github.com/oasisprotocol/oasis-indexer/common"
+	"github.com/oasisprotocol/oasis-indexer/config"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+	"github.com/oasisprotocol/oasis-indexer/storage/client"
+)
+
+const evmContractsVerifierName = "evmcontracts-verifier"
+
+var evmRuntimes = []common.Runtime{common.RuntimeEmerald, common.RuntimeSapphire}
+
+// evmContractsVerifier is an analyzer that verifies EVM contracts.
+type evmContractsVerifier struct {
+	interval time.Duration
+	target   storage.TargetStorage
+	logger   *log.Logger
+	chain    common.ChainName
+	client   *sourcify.SourcifyClient
+}
+
+var _ analyzer.Analyzer = (*evmContractsVerifier)(nil)
+
+func (a *evmContractsVerifier) Name() string {
+	return evmContractsVerifierName
+}
+
+type unverifiedContract struct {
+	Addr                  string
+	AddrContextIdentifier string
+	AddrContextVersion    int
+	AddrData              []byte
+	EthAddress            ethCommon.Address
+}
+
+func (a *evmContractsVerifier) getUnverifiedContracts(ctx context.Context, runtime common.Runtime) ([]*unverifiedContract, error) {
+	var unverifiedContracts []*unverifiedContract
+	rows, err := a.target.Query(ctx, queries.RuntimeEVMUnverfiedContracts, runtime)
+	if err != nil {
+		return nil, fmt.Errorf("querying unverified contracts: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var contract unverifiedContract
+		if err = rows.Scan(
+			&contract.Addr,
+			&contract.AddrContextIdentifier,
+			&contract.AddrContextVersion,
+			&contract.AddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning unverified contracts: %w", err)
+		}
+
+		// Compute the eth address from the preimage.
+		ethAddress, err := client.EVMEthAddrFromPreimage(contract.AddrContextIdentifier, contract.AddrContextVersion, contract.AddrData)
+		if err != nil {
+			return nil, fmt.Errorf("contract eth address: %w", err)
+		}
+		// TODO: assert eth address 20 bytes?
```

Comment:
It's fine either way. Probably not particularly needed; a lot of things will start going wrong if this is not 20 bytes, so we'll notice fast. We do verify it in some other places IIRC, mostly implicitly by using `go-ethereum` to hexify.